internal/utils/status/status.go

```diff
@@ -0,0 +1,117 @@
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package status
+
+import (
+	"context"
+	"fmt"
+
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/client-go/dynamic"
+)
+
+// ConditionsFromUnstructured fetches all of the status.conditions from an
+// unstructured object and returns them as a slice of metav1.Condition.  The
+// status.conditions field is expected to be a slice of map[string]any
+// which can be cast into a metav1.Condition.
+func ConditionsFromUnstructured(unstrObj *unstructured.Unstructured) ([]metav1.Condition, error) {
+	objKind, objName := ObjKindName(unstrObj)
+
+	// Iterate the status conditions and ensure each condition reports a "Ready"
+	// status.
+	unstrConditions, found, err := unstructured.NestedSlice(unstrObj.Object, "status", "conditions")
+	if !found {
+		return nil, fmt.Errorf("no status conditions found for %s: %s", objKind, objName)
+	}
+	if err != nil {
+		return nil, fmt.Errorf("failed to get status conditions for %s: %s: %w", objKind, objName, err)
+	}
+
+	conditions := make([]metav1.Condition, 0, len(unstrConditions))
+
+	for _, condition := range unstrConditions {
+		conditionMap, ok := condition.(map[string]any)
+		if !ok {
+			return nil, fmt.Errorf("expected %s: %s condition to be type map[string]any, got: %T",
+				objKind, objName, conditionMap)
+		}
+
+		var c *metav1.Condition
+
+		if err := runtime.DefaultUnstructuredConverter.FromUnstructured(conditionMap, &c); err != nil {
+			return nil, fmt.Errorf("failed to convert condition map to metav1.Condition: %w", err)
+		}
+
+		conditions = append(conditions, *c)
+	}
+
+	return conditions, nil
+}
+
+type ResourceNotFoundError struct {
+	Resource string
+}
+
+func (e ResourceNotFoundError) Error() string {
+	return fmt.Sprintf("no %s found, ignoring since object must be deleted or not yet created", e.Resource)
+}
+
+// ConditionsFromResource fetches the conditions from a resource identified by
+// the provided GroupVersionResource and labelSelector.  The function returns
+// the name and kind of the resource and a slice of metav1.Condition.  If the
+// resource is not found, returns a ResourceNotFoundError which can be checked
+// by the caller to prevent reconciliation loops.
+//
+//nolint:revive
+func ConditionsFromResource(
+	ctx context.Context, dynamicClient dynamic.Interface,
+	gvr schema.GroupVersionResource, labelSelector string,
+) (string, string, []metav1.Condition, error) {
```

Comment:
Function signature looks confusing to me tbh. Structs in the arguments and in the return statement would be much clearer. 
As an option the list op could be done at `checkProviderStatus` and here you can do only basic checks on `metav1.List`

Reply:
@zerospiel had some ideas of how to make this easier but I've rebased this damn thing so many times I rebased my changes away apparently.  I don't mind doing a struct.

Reply:
I did a struct for the return fields which helped alleviate some of the confusion with types that match.  

>> As an option the list op could be done at checkProviderStatus and here you can do only basic checks on metav1.List

A lot of the choices made were so that we could use similar functions in multiple places instead of something specific to `checkProviderStatus`.

The `ConditionsFromResource` function is performing the list, the condition parsing is within `ConditionsFromUnstructured`, so perhaps the name here could change to make it less confusing.  A modified `ConditionsFromResource` to `GetResourceConditions`